### PR TITLE
Enhancements to VObject

### LINKED ISSUE #1: Python 3 compatibility
Installation fails on Python 3:

```
$ pip install vobject
Collecting vobject
  Downloading vobject-0.6.6.tar.gz (53kB)
    100% |████████████████████████████████| 57kB 3.0MB/s 
    Complete output from command python setup.py egg_info:
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "/tmp/pip-build-rcah3vyp/vobject/setup.py", line 31, in <module>
        from ez_setup import use_setuptools
      File "/tmp/pip-build-rcah3vyp/vobject/ez_setup/__init__.py", line 94
        except pkg_resources.VersionConflict, e:
                                            ^
    SyntaxError: invalid syntax
```

### PR DESCRIPTION
**Pull request by skarim**
@tBaxter's enhancements from his [fork](https://github.com/tBaxter/vobject)

- Add Python 3 compatibility (fixes #1)
- Updated version of `python-dateutil` (now requires version 2.4.0)
- More comprehensive unit tests
- Minor performance improvements
- (PyPI only) Test files are included in download package

### CODE REVIEW

File: vobject/vcard.py [hunk 1]
```diff
@@ -287,7 +312,7 @@ def transformToNative(obj):
         """Turn obj.value into a list."""
         if obj.isNative: return obj
         obj.isNative = True
-        obj.value = splitFields(obj.value)
+        # obj.value = splitFields(obj.value)
```
Comment (skarim) [inline vobject/vcard.py]: remove or uncomment

File: vobject/vcard.py [hunk 2]
```diff
@@ -219,21 +234,25 @@ def toList(stringOrList):
 
 def serializeFields(obj, order=None):
     """Turn an object's fields into a ';' and ',' seperated string.
-    
+
     If order is None, obj should be a list, backslash escape each field and
     return a ';' separated string.
     """
     fields = []
+    print("Inside serializeFields")
```
Comment (skarim) [inline vobject/vcard.py]: extraneous print statement

File: vobject/change_tz.py [hunk 3]
```diff
@@ -70,9 +85,9 @@ def get_options():
 
     (cmdline_options, args) = parser.parse_args()
     if not args and not cmdline_options.list:
-        print "error: too few arguments given"
-        print
-        print parser.format_help()
+        print("error: too few arguments given")
+        print()
```
Comment (skarim) [inline vobject/change_tz.py]: should be just `print` for Python 2 support

File: vobject/base.py [hunk 4]
```diff
@@ -991,27 +1009,22 @@ def modifyTop(self, item):
...
-    """Generate one Component at a time from a stream.
-
-    >>> import StringIO
-    >>> f = StringIO.StringIO(testVCalendar)
-    >>> cal=readComponents(f).next()
-    >>> cal
-    <VCALENDAR| [<VEVENT| [<SUMMARY{u'BLAH': [u'hi!']}Bastille Day Party>]>]>
-    >>> cal.vevent.summary
-    <SUMMARY{u'BLAH': [u'hi!']}Bastille Day Party>
-
```
Comment (skarim) [inline vobject/base.py]: docstrings deleted, need to re-add

File: vobject/base.py [hunk 5]
```diff
@@ -941,37 +960,36 @@ def defaultSerialize(obj, buf, lineLength):
...
+            foldOneLine(outbuf, "{0}END:{1}".format(groupString, obj.name), lineLength)
+
     elif isinstance(obj, ContentLine):
         startedEncoded = obj.encoded
-        if obj.behavior and not startedEncoded: obj.behavior.encode(obj)
-        s=codecs.getwriter('utf-8')(cStringIO.StringIO()) #unfolded buffer
+        if obj.behavior and not startedEncoded:
+            obj.behavior.encode(obj)
+
+        #s = codecs.getwriter('utf-8')(six.StringIO()) #unfolded buffer
```
Comment (skarim) [inline vobject/base.py]: remove or uncomment

File: vobject/base.py [hunk 6]
```diff
@@ -941,37 +960,36 @@ def defaultSerialize(obj, buf, lineLength):
         else:
             groupString = obj.group + '.'
         if obj.useBegin:
-            foldOneLine(outbuf, str(groupString + u"BEGIN:" + obj.name), lineLength)
+            foldOneLine(outbuf, "{0}BEGIN:{1}".format(groupString, obj.name), lineLength)
         for child in obj.getSortedChildren():
-            #validate is recursive, we only need to validate once
+            # validate is recursive, we only need to validate once
             child.serialize(outbuf, lineLength, validate=False)
+            # print('child serialized', str(child))
```
Comment (skarim) [inline vobject/base.py]: need to remove

File: vobject/base.py [hunk 7]
```diff
@@ -915,25 +918,41 @@ def foldOneLine(outbuf, input, lineLength = 75):
...
                 written = len(input)
             else:
                 # Check whether next char is valid utf8 lead byte
-                while (input[offset] > 0x7F) and ((ord(input[offset]) & 0xC0) == 0x80):
-                    # Step back until we have a valid char
-                    offset -= 1
-                
+                # while (input[offset] > 0x7F) and ((ord(input[offset]) & 0xC0) == 0x80):
+                #    # Step back until we have a valid char
+                #    offset -= 1
```
Comment (skarim) [inline vobject/base.py]: remove or uncomment

File: vobject/base.py [hunk 8]
```diff
@@ -809,28 +805,26 @@ def getLogicalLines(fp, allowQP=True, findBegin=False):
...
-                    raise ParseError, 'Could not find BEGIN when trying to determine encoding'
-        else:
-            val = bytes
-        
+                    raise ParseError('Could not find BEGIN when trying to determine encoding')
+        """
         # strip off any UTF8 BOMs which Python's UTF8 decoder leaves
-
-        val = val.lstrip( unicode( codecs.BOM_UTF8, "utf8" ) )
+        #val = val.lstrip( unicode( codecs.BOM_UTF8, "utf8" ) )
```
Comment (skarim) [inline vobject/base.py]: remove or uncomment

File: vobject/base.py [hunk 9]
```diff
@@ -809,28 +805,26 @@ def getLogicalLines(fp, allowQP=True, findBegin=False):
...
                             break
                     except UnicodeDecodeError:
                         pass
                 else:
-                    raise ParseError, 'Could not find BEGIN when trying to determine encoding'
-        else:
-            val = bytes
-        
+                    raise ParseError('Could not find BEGIN when trying to determine encoding')
+        """
```
Comment (skarim) [inline vobject/base.py]: remove or uncomment

File: vobject/base.py [hunk 10]
```diff
@@ -734,30 +747,11 @@ def parseParams(string):
...
-    >>> parseLine('EMAIL;TYPE="blah",hah;INTERNET="DIGI",DERIDOO:john@nowhere.com')
-    ('EMAIL', [['TYPE', 'blah', 'hah'], ['INTERNET', 'DIGI', 'DERIDOO']], 'john@nowhere.com', None)
-    >>> parseLine('item1.ADR;type=HOME;type=pref:;;Reeperbahn 116;Hamburg;;20359;')
-    ('ADR', [['type', 'HOME'], ['type', 'pref']], ';;Reeperbahn 116;Hamburg;;20359;', 'item1')
-    >>> parseLine(":")
-    Traceback (most recent call last):
-    ...
-    ParseError: 'Failed to parse line: :'
-    """
-
```
Comment (skarim) [inline vobject/base.py]: docstrings deleted, need to re-add

File: vobject/base.py [hunk 11]
```diff
@@ -711,14 +732,6 @@ class NativeError(VObjectError):
...
 
 def parseParams(string):
-    """
-    >>> parseParams(';ALTREP="http://www.wiz.org"')
-    [['ALTREP', 'http://www.wiz.org']]
-    >>> parseParams('')
-    []
-    >>> parseParams(';ALTREP="http://www.wiz.org;;",Blah,Foo;NEXT=Nope;BAR')
-    [['ALTREP', 'http://www.wiz.org;;', 'Blah', 'Foo'], ['NEXT', 'Nope'], ['BAR']]
-    """
```
Comment (skarim) [inline vobject/base.py]: docstrings deleted, need to re-add

File: vobject/base.py [hunk 12]
```diff
@@ -259,18 +296,7 @@ def updateTable(x):
...
-        # self.value should be unicode for iCalendar, but if quoted-printable
-        # is used, or if the quoted-printable state machine is used, text may be
-        # encoded
-        if type(self.value) is str:
-            charset = 'iso-8859-1'
-            if 'CHARSET' in self.params:
-                charsets = self.params.pop('CHARSET')
-                if charsets:
-                    charset = charsets[0]
-            self.value = unicode(self.value, charset)
```
Comment (skarim) [inline vobject/base.py]: support for quoted-printable was removed

File: test_files/more_tests.txt [hunk 13]
```diff
@@ -68,19 +68,3 @@ Ruby escapes semi-colons in rrules
...
-<Name:  ?   >
->>> vcf.n.value.given
-u'\xe9'
->>> vcf.serialize()
-'BEGIN:VCARD\r\nVERSION:2.1\r\nFN:\xc3\xa9\r\nN:;\xc3\xa9;;;\r\nTEL:0111111111\r\nEND:VCARD\r\n'
-
->>> vcs = 'BEGIN:VCALENDAR\r\nPRODID:-//OpenSync//NONSGML OpenSync vformat 0.3//EN\r\nVERSION:1.0\r\nBEGIN:VEVENT\r\nDESCRIPTION;CHARSET=UTF-8;ENCODING=QUOTED-PRINTABLE:foo =C3=A5=0Abar =C3=A4=\r\n=0Abaz =C3=B6\r\nUID:20080406T152030Z-7822\r\nEND:VEVENT\r\nEND:VCALENDAR\r\n'
->>> vcs = vobject.readOne(vcs, allowQP = True)
->>> vcs.serialize()
-'BEGIN:VCALENDAR\r\nVERSION:1.0\r\nPRODID:-//OpenSync//NONSGML OpenSync vformat 0.3//EN\r\nBEGIN:VEVENT\r\nUID:20080406T152030Z-7822\r\nDESCRIPTION:foo \xc3\xa5\\nbar \xc3\xa4\\nbaz \xc3\xb6\r\nEND:VEVENT\r\nEND:VCALENDAR\r\n'
```
Comment (skarim) [inline test_files/more_tests.txt]: tests for quoted-printable were removed

**Comment by pypingou** [comment]
While there is legally nothing preventing you to do this, I hope this has been discussed with @tBaxter since otherwise we'll end up with 2 modules on pypi with the exact same code :]

**Comment by skarim** [comment]
Yup I've been corresponding with @tBaxter to make sure everyone is on board :+1: 

There's still a large number of people using the original VObject on PyPi, so I think it will benefit everyone if these improvements are released :)

**Comment by tBaxter** [comment]
Not super wild about it, but whatever.

**Comment by skarim** [comment]
Of course, my bad for missing it earlier!
Updated: https://github.com/eventable/vobject/blob/master/ACKNOWLEDGEMENTS.txt

**Pull request merged**